init : witness To SMT

0xsharma opened the pull request:
In this PR, we create method to create an SMT from a witness.

0xsharma committed:
init : witness To SMT

revitteth commented:
Hey @0xsharma please could you comment with something like `I have read the Individual/Corporate CLA document and accept its terms` and I will summon the CLA bot to note your acceptance so we can in future merge this!

0xsharma committed:
add : insertHashNode function

0xsharma commented:
I have read the Individual/Corporate CLA document and accept its terms

0xsharma committed:
Merge branch '0xPolygonHermez:zkevm' into shivam/POS-2445

cffls committed:
Fix case where hash nodes present in witness

cffls committed:
Merge remote-tracking branch 'origin/zkevm' into shivam/POS-2445

cffls committed:
New implementation of insert hash node

This commit re-implements hash node insertion in a new function and fixed the failed unit test of "TestWitnessToSMT".

0xsharma committed:
init : basic stateReader

0xsharma committed:
add : stateReader interface methods

cffls committed:
Fix state reader

0xsharma committed:
chg : finalise SMT state reader

0xsharma committed:
Merge branch 'zkevm' into shivam/POS-2445-upstream

0xsharma committed:
Merge pull request #2 from 0xsharma/shivam/POS-2445-upstream

Merge "zkevm" branch into shivam/POS-2445

0xsharma committed:
chg : SMT ReadAccountStorage

cffls commented on smt/pkg/smt/smt_state_reader.go:
```diff
@@ -0,0 +1,245 @@
+}
+
+// getStorageMap returns the storage map of an address from the SMT
+func (s *SMT) getStorageMap(address libcommon.Address) (map[libcommon.Hash]libcommon.Hash, error) {
```
This function traverses the entire trie every time. It could be optimized a bit by storing the map in memory after the first traversal.

0xsharma replied:
Done ✅

cffls commented on smt/pkg/smt/witness_test.go:
```diff
@@ -101,6 +103,90 @@ func findNode(t *testing.T, w *trie.Witness, addr libcommon.Address, storageKey
+	newAccCode, _ := newSMT.ReadAccountCode(contract, 0, newAcc.CodeHash)
+	expectedAccCodeSize, _ := smtTrie.ReadAccountCodeSize(contract, 0, expectedAcc.CodeHash)
+	newAccCodeSize, _ := newSMT.ReadAccountCodeSize(contract, 0, newAcc.CodeHash)
+
```
It will be great if we can test retrieving the value from a storage slot as well.

0xsharma replied:
Done ✅

0xsharma committed:
fix : stateReader storage reads

cffls reviewed (approved):
LGTM, thanks!

0xsharma committed:
Merge branch 'zkevm' into shivam/POS-2445

0xsharma committed:
Merge branch 'zkevm' into shivam/POS-2445

cffls commented:
@hexoscott @revitteth could you take a look at this? This will enable us to replay and verify a block directly from a witness in erigon.

hexoscott commented:
Looks good, we just need to ensure that it doesn't affect the usual witness generation as we've made tweaks there recently to support the executor / prover

mandrigin commented:
closing stale PR
